### PR TITLE
Feature/spark3 tests support

### LINKED ISSUE #128: Add automated tests for Spark integration with Spark 3.x in addition to Spark 2.4

### PR DESCRIPTION
**Pull request by OleksandrDvornik**
Resolves #128 
Adding new test data for spark3.
Solving few issues on spark3 migration

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -223,12 +223,41 @@ jobs:
...
       - store_artifacts:
-          path: integration/spark/build/reports/tests/integrationTests
+          path: integration/spark/build/reports/tests/integrationTest
+          destination: test-report
+      - save_cache:
+          key: v1-integration-spark-{{ .Branch }}-{{ .Revision }}
+          paths:
+            - ~/.gradle
+
+  test-integration-spark-v3:
```
Comment (wslulciuc) [inline .circleci/config.yml:238]: Can we define a [matrix job](https://circleci.com/blog/circleci-matrix-jobs) to handle different spark versions?

File: integration/spark/build.gradle [hunk 2]
```diff
@@ -104,6 +102,20 @@ dependencies {
...
+        runtimeOnly "com.fasterxml.jackson.module:jackson-module-scala_2.12:${jacksonVersion}"
+        testRuntimeOnly 'com.google.cloud.spark:spark-bigquery-with-dependencies_2.12:0.21.1'
+        testRuntimeOnly "org.apache.spark:spark-core_2.12:${spark3Version}"
+        testRuntimeOnly "org.apache.spark:spark-sql_2.12:${spark3Version}"
+    } else {
+        runtimeOnly "com.fasterxml.jackson.module:jackson-module-scala_2.11:${jacksonVersion}"
+        testRuntimeOnly 'com.google.cloud.spark:spark-bigquery-with-dependencies_2.11:0.21.1'
+        testRuntimeOnly "org.apache.spark:spark-core_2.11:${sparkVersion}"
+        testRuntimeOnly "org.apache.spark:spark-sql_2.11:${sparkVersion}"
+    }
```
Comment (collado-mike) [inline integration/spark/build.gradle:118]: We should be able to do this with Gradle build `configurations`.  See https://docs.gradle.org/6.8.1/userguide/java_testing.html#sec:configuring_java_integration_tests

Define a `configurations` block like
```
configurations {
    spark3.extendsFrom compileOnly
    spark2.extendsFrom compileOnly
    spark3Test.extendsFrom testImplementation
    spark2Test.extendsFrom testImplementation
}
```
Configure each of the dependencies with the appropriate configuration label. Then, in the `test` block, you can update the classpath with the appropriate configuration's classpath:

```
test {
    useJUnitPlatform {
        excludeTags 'integration-test'
    }
    classpath = configurations.spark3.runtimeClasspath
    // ...
```

This gets us closer to running compilation in multiple steps with different sets of dependencies.

File: integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/$VERSION/integration/spark",
+          "name": "file",
+          "uri": "file",
+          "_schemaURL": "https://openlineage.io/spec/1-0-1/OpenLineage.json#/definitions/DatasourceDatasetFacet"
+        },
+        "stats" : {
+          "_producer" : "https://github.com/OpenLineage/OpenLineage/tree/$VERSION/integration/spark",
+          "_schemaURL" : "https://github.com/OpenLineage/OpenLineage/tree/$VERSION/integration/spark/facets/spark-2.4/v1/output-statistics-facet.json",
+          "rowCount" : 0,
+          "size" : -1,
```
Comment (collado-mike) [inline integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json:68]: Are the output stats the only things that differ between the Spark 2 and Spark 3 output? Can we just ignore or find/replace these values in the string content before we do the comparison? Similar to how we find/replace the schema version or ignore the run_id in other tests?

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java [hunk 4]
```diff
@@ -108,6 +116,15 @@ public void setActiveJob(ActiveJob activeJob) {
...
   }
 
+  private Field getConfigField(Function2<TaskContext, Iterator<?>, ?> fn)
+      throws NoSuchFieldException {
+    try {
+      return fn.getClass().getDeclaredField("config$1");
+    } catch (NoSuchFieldException e) {
+      return fn.getClass().getDeclaredField("arg$1");
+    }
+  }
```
Comment (collado-mike) [inline integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java:126]: A comment here would be super helpful

Comment (OleksandrDvornik) [reply integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java:126]: you mean in code? I'll add it

File: integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json (hunk 3 above)
Comment (OleksandrDvornik) [reply integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json:68]: there are few additional fields, but the problem with `mockserver` matcher, I'll take a look can we somehow ignore fields there

Comment (OleksandrDvornik) [reply integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json:68]:
```
"schema": [
              {
                "name": "count",
                "dataType": {
                  "numeric": {},
                  "integral": {},
                  "ordering": {}
                },
                "nullable": false,
                "metadata": {
                  "map": {}
                },
                "comment": null
              }
            ],
```

vs
```

 "schema": {
              "type": "struct",
              "fields": [
                {
                  "name": "count",
                  "type": "long",
                  "nullable": false,
                  "metadata": {}
                }
              ]
            },
```

File: integration/spark/build.gradle (hunk 2 above)
Comment (OleksandrDvornik) [reply integration/spark/build.gradle:118]: so, you want to have separate test tasks
for spark 2.4.7
```
test {
...
}
integrationTest {
}

and for spark 3
```
testSpark3 {
...
}
integrationTestSpark3 {
}
```
Cause I wanted to have only two of them
```
test {
...
}
integrationTest {
}
```
and in CI add param to build it for spark3

File: integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json (hunk 3 above)
Comment (OleksandrDvornik) [reply integration/spark/integrations/spark3/container/pysparkRddToTableCompleteEvent.json:68]: we can use placeholders in mock server
https://www.mock-server.com/mock_server/creating_expectations.html#button_match_request_by_body_with_json_placeholders

File: .circleci/config.yml (hunk 1 above)
Comment (OleksandrDvornik) [reply .circleci/config.yml:238]: nice, I guess we can do it.

File: integration/spark/build.gradle (hunk 2 above)
Comment (OleksandrDvornik) [reply integration/spark/build.gradle:115]: @collado-mike what do you think if we leave only two tasks and in CI we will use matrix-job, as @wslulciuc  suggests.
If we want to have separate tasks, we can use it as:

```
- run: ./gradlew --no-daemon --stacktrace << parameters.gradle-task >>

...
parameters:
              gradle-task: [test, testSpark3]
```
vs
```
- run: ./gradlew --no-daemon --stacktrace test -Pspark.version=<< parameters.spark-version >>
...
parameters:
     spark-version: [2.4.7, 3.1]
```

**Comment by OleksandrDvornik** [comment]
I've remove duplicate test files for spark3. Use configurations instead of if block.

File: integration/spark/build.gradle [hunk 5]
```diff
@@ -116,10 +135,16 @@ task javadocJar(type: Jar, dependsOn: javadoc) {
...
-    useJUnitPlatform {
-        excludeTags 'integration-test'
+task bumpVersionTestFiles() {
+    doLast {
+        def dir = new File(project.projectDir.path + '/integrations')
+        dir.eachFileRecurse (FileType.FILES) { file ->
+            ant.replace(file: file, token: "\$VERSION", value: project.version.toString())
+        }
     }
+}
```
Comment (collado-mike) [inline integration/spark/build.gradle:145]: Please let's not do this. I removed this functionality before because it clutters the git workspace- every time I have to remember to restore these files after running the build. Let's just do the find/replace in the tests when we read these files.

File: integration/spark/build.gradle [hunk 6]
```diff
@@ -128,25 +153,27 @@ test {
     }
     systemProperties = [
             'junit.platform.output.capture.stdout': 'true',
-            'junit.platform.output.capture.stderr': 'true'
+            'junit.platform.output.capture.stderr': 'true',
+            'spark.version'                       : project.getProperty('spark.version'),
+            'openlineage.spark.jar': "${archivesBaseName}-${project.version}.jar"
     ]
+    classpath =  project.sourceSets.test.runtimeClasspath + (isSpark3 ? configurations.spark3Test : configurations.spark2Test)
```
Comment (collado-mike) [inline integration/spark/build.gradle:160]: If you update the configurations to extend `testImplementation`, you don't have to do this concatenation. E.g., see the docs at https://docs.gradle.org/current/userguide/declaring_dependencies.html:

```groovy
configurations {
    smokeTest.extendsFrom testImplementation
}
```

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java [hunk 7]
```diff
@@ -45,7 +45,9 @@ public static LogicalPlanSerializer getInstance() {
   private LogicalPlanSerializer() {
     mapper = new ObjectMapper();
     try {
-      mapper.registerModule(DefaultScalaModule$.MODULE$);
+      mapper.registerModule(
+          (Module)
+              Class.forName("com.fasterxml.jackson.module.scala.DefaultScalaModule").newInstance());
```
Comment (collado-mike) [inline integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java:50]: These class names are different. Are we sure this works in both Spark versions?

**Review by collado-mike — COMMENTED** [review]
Looks good! Can we just fix the build config and stop altering files in the git workspace?

File: integration/spark/build.gradle (hunk 6 above)
Comment (OleksandrDvornik) [reply integration/spark/build.gradle:160]: I guess it's more complex. It was not working for me.
Receiving a lot of ClassNotFoundException
`Caused by: java.lang.ClassNotFoundException: com.google.cloud.spark.bigquery.repackaged.com.google.cloud.bigquery.MockBigQueryRelationProvider$MockInjector
...
Caused by: java.lang.ClassNotFoundException: io.openlineage.spark.agent.lifecycle.plan.JDBCRelationVisitorTest
...
`
For 
```
spark3Test.extendsFrom testRuntimeClasspath
spark2Test.extendsFrom testRuntimeClasspath
```
and
```
spark3Test.extendsFrom testImplementation
spark2Test.extendsFrom testImplementation
```

and usage
`classpath = isSpark3 ? configurations.spark3Test : configurations.spark2Test`

`classpath = configurations.spark3.runtimeClasspath` fails with
> Could not get unknown property 'runtimeClasspath' for configuration ':spark2Test'

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java (hunk 7 above)
Comment (OleksandrDvornik) [reply integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java:50]: It's not working for spark3, due to `relocate` jackson libs.
Even if I get a `DefaultScalaModule$` from 2.12 , then If I try to register it  would receive an `ClassCastExeption`, that `com.fasterxml...Module `can.'t be cast to `openlineage.com.fasterxml...Module` That's why we need to have [PR](https://github.com/OpenLineage/OpenLineage/pull/275)

File: integration/spark/build.gradle [hunk 8]
```diff
@@ -78,16 +85,12 @@ dependencies {
     implementation "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
     implementation "com.fasterxml.jackson.core:jackson-core:${jacksonVersion}"
     implementation "com.fasterxml.jackson.datatype:jackson-datatype-jsr310:${jacksonVersion}"
-    implementation "com.fasterxml.jackson.module:jackson-module-scala_2.11:${jacksonVersion}"
+    compileOnly "com.fasterxml.jackson.module:jackson-module-scala_2.11:${jacksonVersion}"
```
Comment (OleksandrDvornik) [inline integration/spark/build.gradle:88]: `LogicalPlanSerializer` wouldn't  work with `compileOnly jackson-module-scala_2.11`  for 2.4.7 and for 3.1. Due to `relocate` in `shadowJar`. In compile time we have openlineage prefix, and if we bound it to our jar, it would work fine for 2.4.7. If we run it for spark3, we would get   
```
java.lang.NoSuchMethodError: scala.Product.$init$(Lscala/Product;)V
      at io.openlineage.spark.agent.lifecycle.plan.LogicalRDDVisitorTest.testApply(LogicalRDDVisitorTest.java:49)
```
Cause it loads scala 2.11 compiled classes in 2.12 env

File: integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java [hunk 9]
```diff
@@ -36,16 +36,15 @@
...
-  public static LogicalPlanSerializer getInstance() {
-    return instance;
-  }
-
-  private LogicalPlanSerializer() {
+  public LogicalPlanSerializer() {
     mapper = new ObjectMapper();
     try {
-      mapper.registerModule(DefaultScalaModule$.MODULE$);
+      mapper.registerModule(
```
Comment (OleksandrDvornik) [inline integration/spark/src/main/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializer.java:43]: this code tries to load class dynamically, but it would always fail, cause at runtime we have unprefixed classname
`com.fasterxml.jackson.module.scala.DefaultScalaModule$`
vs expected
`openlineage.com.fasterxml.jackson.module.scala.DefaultScalaModule$`

File: integration/spark/build.gradle (hunk 8 above)
Comment (collado-mike) [reply integration/spark/build.gradle:88]: Are you saying that `compileOnly` jars are included in the bundled shadow jar? Exclude the jar from the bundle and get rid of the `relocate` statement. The goal is to just use the Jackson dependencies as they are provided by Spark. No renaming,  no bundling.

Comment (OleksandrDvornik) [reply integration/spark/build.gradle:88]: `compileOnly` is not included, but we don't have it in runtime (with prefix), that's why it wouldn't work. Removing `relocate` in in other PR. That's what I meant when say, that this PR relates to [this](https://github.com/OpenLineage/OpenLineage/pull/275).
It should start work, when [PR](https://github.com/OpenLineage/OpenLineage/pull/275) will go live

Comment (collado-mike) [reply integration/spark/build.gradle:88]: Ok, can you rebase one PR on the other one? We shouldn't know that things will work only after everything is merged. It seems to me the Jackson PR should go first, then the Spark 3 tests can be rebased on top of that one. Then we should be able to validate everything works the way we expect.

Comment (OleksandrDvornik) [reply integration/spark/build.gradle:88]: yes, Jackson PR should go first. I can rebase it when we merge that PR.
Second option I see is to create a separate branch `spark3` and change PR base to it. With that approach we would have a small separate PRs as we discussed previously, and it wouldn't impact `main` and changes in `main` wouldn't impact spark3 support. Later when Jackson PR, this one and Source Tree PRs will be merged, we can pull changes from main, test them and merge to main.

Comment (OleksandrDvornik) [reply integration/spark/build.gradle:88]: I would prefer to have both PR's to start working on Splitting Source Tree Issue. Cause it can cause the same issue as right now.

Comment (collado-mike) [reply integration/spark/build.gradle:88]: You can simply rebase the `feature/spark3_tests_support` on the `feature/jackson-serialization-test-for-using-provided-libs` branch and change this PR to use the other branch as its base. Then we can see both sets of changes and you'll be able to verify that they both work. We don't have to wait until the other change is merged to do that.

Unfortunately, we're splitting conversations about dependencies and serialization between both of these PRs and it's hard to follow what's going on.

Comment (OleksandrDvornik) [reply integration/spark/build.gradle:88]: ok

File: integration/spark/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java [hunk 10]
```diff
@@ -230,6 +210,42 @@ public void testSerializeInsertIntoHadoopPlan() throws IOException {
     assertThat(hadoopFSActualNode).satisfies(new MatchesMapRecursively(expectedHadoopFSNode));
   }
 
+  private CatalogTable getCatalogTable() throws InvocationTargetException, IllegalAccessException {
+    Method applyMethod =
+        Arrays.stream(CatalogTable.class.getDeclaredMethods())
+            .filter(m -> m.getName().equals("apply"))
+            .findFirst()
+            .get();
```
Comment (collado-mike) [inline integration/spark/src/test/java/io/openlineage/spark/agent/lifecycle/LogicalPlanSerializerTest.java:218]: Yuck! Let's add a comment to address this when we split the source tree

**Review by collado-mike — APPROVED** [review]
Awesome work! Glad to see this finally getting in!
